File: Ejemplos/05-ValorReferencia.java
class ValorReferencia {

    public static void main(String[] args) {
        int num1 = 10;
        int num2 = 20;
        
        int res = sumar(num1, num2);

        System.out.println("El resultado de la suma " + num1 + " + " + num2 + " = " + res);
        int numero = 10;
        res = pasoValor(numero);
        System.out.println("3.- paso por valor: " + res);
        System.out.println(numero);

        int[] numeros = new int[1];
        numeros[0] = 10;
        System.out.println(numeros[0]);
        res = pasoReferencia(numeros);
        System.out.println("3.- paso por valor: " + res);
        System.out.println(numeros[0]);
    }

    public static int sumar(int a, int b) {
      a = 200;
      System.out.println("Funcion sumar");
      System.out.println("a: " + a);
      System.out.println("b: " + b);
      return a + b;
    }

    public static int pasoValor(int num) {
      System.out.println("Paso valor");
      // int res = num * 2;
      // return res;
      
      // num = num * 2;
      // return num;

      // Esto sí deja, si queremos modificar el valor debemos copiarlo a una variable local
      // val res = num * 2
      // return res

      // Esto también
      return num * 2;
  }

  public static int pasoReferencia(int[] num) {
    System.out.println("Paso referencia");
    num[0] = num[0] + 2;
    return num[0];
  }
}

File: Ejemplos/05-Valor-referencia.kt
/**
 * Programa o función principal
 * Siempre tiene que existir al menos 1
 */
fun main(){
  // Procedimiento
  var num1 = 10
  var num2 = 20
  var res = sumar(num1, num2)
  println("1.- La suma de $num1 + $num2 = $res")

  res = sumar(10, 20)
  println("2.- La suma es: $res")

  val numero = 10
  res = pasoValor(numero)
  println("3.- paso por valor: $res")

  val numeros = intArrayOf(10)
   println(numeros[0])
  res = pasoReferencia(numeros)
  println("3.- paso por valor: $res")
  println(numeros[0])

}

fun sumar(a: Int, b: Int): Int{
  println("Funcion sumar")
  println("a: $a")
  println("b: $b")
  return a + b
}

fun pasoValor(num: Int): Int {
  println("Paso valor")
  //return num * 2
  // val res = num * 2
  // return num
  // No se puede hacer porque se bloquea por el compilador de Kotlin
  // Al ser un paso por valor
  // num = num * 2
  // return num

  // Esto sí deja, si queremos modificar el valor debemos copiarlo a una variable local
  // val res = num * 2
  // return res

  // Esto también
  return num * 2
}

fun pasoReferencia(num: IntArray): Int {
  println("Paso referencia")
  num[0] = num[0] + 2
  return num[0]
}
  
  